src/morx.rs

```diff
@@ -0,0 +1,1824 @@
+                return 2;
+            } else {
+                return val;
+            }
+        }
+    }
+}
+
+//--------------------------------- Contextual Substitution ---------------------------------------------
+pub struct ContextualSubstitution<'a> {
+    glyphs: &'a mut Vec<RawGlyph<()>>,
+    next_state: u16,
+    mark: Option<(usize, u16)>, //record marked glyph and its position: (position, mark_glyph)
+}
+
+impl<'a> ContextualSubstitution<'a> {
+    pub fn new(glyphs: &'a mut Vec<RawGlyph<()>>) -> ContextualSubstitution<'a> {
+        ContextualSubstitution {
+            glyphs: glyphs,
+            next_state: 0,
+            mark: None,
+        }
+    }
+
+    pub fn process_glyphs<'b>(
+        &mut self,
+        contextual_subtable: &ContextualSubtable<'b>,
+    ) -> Result<(), ParseError> {
+        const SET_MARK: u16 = 0x8000;
+        const DONT_ADVANCE: u16 = 0x4000;
+        let mut old_glyph: u16;
+        let mut new_glyph: u16;
+
+        //loop through glyphs:
+        for i in 0..self.glyphs.len() {
+            let current_glyph: u16 = self.glyphs[i].glyph_index;
+            old_glyph = self.glyphs[i].glyph_index;
+            new_glyph = self.glyphs[i].glyph_index;
+
+            let mut class = glyph_class(current_glyph, &contextual_subtable.class_table);
+
+            'glyph: loop {
+                let index_to_entry_table;
+                let entry;
+
+                if let Some(state_row) = contextual_subtable
+                    .state_array
+                    .state_array
+                    .get(usize::from(self.next_state))
+                {
+                    index_to_entry_table = state_row.read_item(usize::from(class))?;
+                } else {
+                    return Err(ParseError::BadIndex);
+                }
+
+                if let Some(contxt_entry) = contextual_subtable
+                    .entry_table
+                    .contextual_entries
+                    .get(usize::from(index_to_entry_table))
+                {
+                    entry = contxt_entry;
+                } else {
+                    return Err(ParseError::BadIndex);
+                }
+
+                self.next_state = entry.next_state;
+
+                //if there is a marked glyph on record and the entry is providing a mark_index to the
+                //substitution table for it, then make the substitution for the marked glyph.
+                if entry.mark_index != 0xFFFF {
+                    if let Some((mark_pos, mark_glyph)) = self.mark {
+                        if let Some(mark_glyph_subst) = lookup(
+                            mark_glyph,
+                            &contextual_subtable.substitution_subtables
+                                [usize::from(entry.mark_index)],
+                        ) {
+                            self.glyphs[mark_pos].glyph_index = mark_glyph_subst;
+                            self.glyphs[mark_pos].glyph_origin = GlyphOrigin::Direct;
```

Comment:
We're directly indexing into things here (using `[ ]`), which makes me nervous as some of the values appear to come from the font data and this operation will panic if the value is out of bounds. I'd suggest using [get](https://doc.rust-lang.org/std/primitive.slice.html#method.get) when the value cannot be guaranteed to be in bounds.